src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,202 @@
+    PUSH_ARGUMENT_REGISTERS
+    // +8 for alignment
+    alloc_stack (SIZEOF_MAX_FP_ARG_SPILL + 8)
+    SAVE_FLOAT_ARGUMENT_REGISTERS 0
+    END_PROLOGUE
+
+    mov rdi, METHODDESC_REGISTER 
+    call C_FUNC(TheUMEntryPrestubWorker)
+    
+    // we're going to tail call to the exec stub that we just setup 
+
+    RESTORE_FLOAT_ARGUMENT_REGISTERS 0
+    free_stack (SIZEOF_MAX_FP_ARG_SPILL + 8)
+    POP_ARGUMENT_REGISTERS
+    TAILJMP_RAX
+
+NESTED_END TheUMEntryPrestub, _TEXT 
+
+//
+// METHODDESC_REGISTER: UMEntryThunk*
+//
+NESTED_ENTRY UMThunkStub, _TEXT, UMThunkStubUnwindFrameChainHandler
+#define UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE (SIZEOF_MAX_INT_ARG_SPILL + SIZEOF_MAX_FP_ARG_SPILL + 0x8)
+#define UMThunkStubAMD64_XMM_SAVE_OFFSET 0x0
+#define UMThunkStubAMD64_INT_ARG_OFFSET (SIZEOF_MAX_FP_ARG_SPILL + 0x8)
+#define UMThunkStubAMD64_METHODDESC_OFFSET SIZEOF_MAX_FP_ARG_SPILL
+
+// {optional stack args passed to callee}   <-- new RSP
+// xmm0                                     <-- RBP
+// xmm1
+// xmm2
+// xmm3
+// xmm4
+// xmm5
+// xmm6
+// xmm7
+// METHODDESC_REGISTER
+// rdi
+// rsi
+// rcx
+// rdx
+// r8
+// r9
+// rbp
+// r12
+// return address                           <-- entry RSP
+        push_nonvol_reg r12
+        push_nonvol_reg rbp                                                                     // stack_args
+        alloc_stack     UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE
+        save_reg_postrsp rdi, (UMThunkStubAMD64_INT_ARG_OFFSET)
+        save_reg_postrsp rsi, (UMThunkStubAMD64_INT_ARG_OFFSET + 0x08)
+        save_reg_postrsp rdx, (UMThunkStubAMD64_INT_ARG_OFFSET + 0x10)
+        save_reg_postrsp rcx, (UMThunkStubAMD64_INT_ARG_OFFSET + 0x18)
+        save_reg_postrsp r8,  (UMThunkStubAMD64_INT_ARG_OFFSET + 0x20)
+        save_reg_postrsp r9,  (UMThunkStubAMD64_INT_ARG_OFFSET + 0x28)
+        save_reg_postrsp METHODDESC_REGISTER, UMThunkStubAMD64_METHODDESC_OFFSET
+        set_frame       rbp, 0                                                                  // stack_args
+        SAVE_FLOAT_ARGUMENT_REGISTERS UMThunkStubAMD64_XMM_SAVE_OFFSET
+        END_PROLOGUE
+
+        //
+        // Call GetThread()
+        //
+        call            C_FUNC(GetThread)
+        test            rax, rax
+        jz              LOCAL_LABEL(DoThreadSetup)
+
+LOCAL_LABEL(HaveThread):
+
+        mov             r12, rax                // r12 <- Thread*
+
+        //
+        // disable preemptive GC
+        //
+        mov             dword ptr [r12 + OFFSETOF__Thread__m_fPreemptiveGCDisabled], 1
+
+        //
+        // catch returning thread here if a GC is in progress
+        //
+        PREPARE_EXTERNAL_VAR g_TrapReturningThreads, rax
+        cmp                  [rax], 0
+        jnz                  LOCAL_LABEL(DoTrapReturningThreadsTHROW)
+
+LOCAL_LABEL(InCooperativeMode):
+
+        restore_reg     METHODDESC_REGISTER, UMThunkStubAMD64_METHODDESC_OFFSET
```

Comment:
This may look better as a direct mov METHODDESC_REGISTER, [rbp + ...] to be consistent with the rest of the function.